pennylane/optimize/lie_gradient.py

```diff
@@ -0,0 +1,352 @@
+                ]
+
+                # loop through all operations on the input tape
+                for op in tape.operations:
+                    for t in tapes_plus + tapes_min:
+                        with t:
+                            qml.apply(op)
+                for i, t in enumerate(tapes_plus):
+                    with t:
+                        qml.PauliRot(
+                            np.pi / 2,
+                            self.lie_algebra_basis_names[i],
+                            wires=list(range(self.nqubits)),
+                        )
+                        for o in obs:
+                            qml.expval(o)
+                for i, t in enumerate(tapes_min):
+                    with t:
+                        qml.PauliRot(
+                            -np.pi / 2,
+                            self.lie_algebra_basis_names[i],
+                            wires=list(range(self.nqubits)),
+                        )
+                        for o in obs:
+                            qml.expval(o)
+                tapes.append((tapes_plus, tapes_min))
+            return tapes, None
+
+        self.algebra_commutator = algebra_commutator
+
+    def step_and_cost(
+        self,
+        *args,
+        **kwargs,
+    ):
+        r"""Update the circuit with one step of the optimizer and return the corresponding
+        objective function value prior to the step.
+
+        Args:
+            *args:
+            **kwargs:
+
+        Returns:
+
+        """
+        omegas = self.get_omegas()
+        non_zero_lie_algebra_elements = []
+        non_zero_omegas = []
+        for i, element in enumerate(omegas):
+            if not np.isclose(element, 0):
+                non_zero_lie_algebra_elements.append(
+                    tuple(
+                        self.pauli_str_to_int[p]
+                        for p in self.lie_algebra_basis_names[i]
+                    )
+                )
+                non_zero_omegas.append(-omegas[i])
+        lie_gradient = qml.Hamiltonian(
+            non_zero_omegas,
+            [self.get_pauli_op(ps) for ps in non_zero_lie_algebra_elements],
+        )
+        temp_circuit = copy.copy(self.circuit)
+
+        @append_time_evolution(lie_gradient, self.stepsize)
+        def circuit(params, wires, **kwargs):
+            """Append the time evolution"""
+            for obj in temp_circuit.qtape.operations:
+                qml.apply(obj)
+
+            return qml.state()
+
+        # TODO: qml.map is not working here, it throws an error for pauli @ pauli objects, which have `name` tensor.
+        # cost_fn = qml.map(circuit, self.observables, device=self.circuit.device, measure='expval')([],[])
+        # cost_fn = qml.QNodeCollection(circuit, self.observables, device=self.circuit.device, measure='expval')([],[])
+        # cost = sum([self.coeffs[i] * obs for i, obs in enumerate(cost_fn)])
+        # print(cost)
```

Comment:
I think this might be avoidable altogether if you allow the optimizer to take the cost function as input, and extract the info you need from it, rather than accepting the constituents and attempting to recreate the cost function?